integrationtests/self/zero_rtt_test.go

```diff
@@ -938,4 +938,119 @@ var _ = Describe("0-RTT", func() {
 		)
 		Expect(restored).To(BeTrue())
 	})
+
+	It("send 0-RTT datagrams", func() {
+		tlsConf := getTLSConfig()
+		clientTLSConf := getTLSClientConfig()
+		dialAndReceiveSessionTicket(tlsConf, getQuicConfig(&quic.Config{
+			EnableDatagrams: true,
+		}), clientTLSConf)
+
+		tracer := newPacketTracer()
+		ln, err := quic.ListenAddrEarly(
+			"localhost:0",
+			tlsConf,
+			getQuicConfig(&quic.Config{
+				Allow0RTT:       true,
+				EnableDatagrams: true,
+				Tracer:          newTracer(tracer),
+			}),
+		)
+		Expect(err).ToNot(HaveOccurred())
+		defer ln.Close()
+		proxy, num0RTTPackets := runCountingProxy(ln.Addr().(*net.UDPAddr).Port)
+		defer proxy.Close()
+
+		// second connection
+		sentMessage := GeneratePRData(100)
+		var receivedMessage []byte
+		received := make(chan struct{})
+		go func() {
+			defer GinkgoRecover()
+			defer close(received)
+			conn, err := ln.Accept(context.Background())
+			Expect(err).ToNot(HaveOccurred())
+			receivedMessage, err = conn.ReceiveMessage(context.Background())
+			Expect(err).ToNot(HaveOccurred())
+			Expect(conn.ConnectionState().Used0RTT).To(BeTrue())
+		}()
+		conn, err := quic.DialAddrEarly(
+			context.Background(),
+			fmt.Sprintf("localhost:%d", proxy.LocalPort()),
+			clientTLSConf,
+			getQuicConfig(&quic.Config{
+				EnableDatagrams: true,
+			}),
+		)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(conn.ConnectionState().SupportsDatagrams).To(BeTrue())
+		Expect(conn.SendMessage(sentMessage)).To(Succeed())
+		<-conn.HandshakeComplete()
+		<-received
+
+		Expect(conn.ConnectionState().Used0RTT).To(BeTrue())
+		Expect(receivedMessage).To(Equal(sentMessage))
+		num0RTT := atomic.LoadUint32(num0RTTPackets)
+		fmt.Fprintf(GinkgoWriter, "Sent %d 0-RTT packets.", num0RTT)
+		Expect(num0RTT).ToNot(BeZero())
+		zeroRTTPackets := get0RTTPackets(tracer.getRcvdLongHeaderPackets())
+		Expect(len(zeroRTTPackets)).To(BeNumerically("==", 1))
+		Expect(conn.CloseWithError(0, "")).To(Succeed())
+	})
+
+	It("reject 0-RTT datagrams when the server doesn't support datagrams anymore", func() {
```

Comment:
Would it make sense to test that the server doesn't process the received DATAGRAM?

Reply:
I thought it could test the check in ValidFor0RTT from a higher level, like test that the tracer will not recevie the 0-RTT packet. But yes, it feels meaningless for a integration test considering datagram can be unreliable. Should I remove it?

Reply:
I'm suggesting to test that the server actually drops the 0-RTT packet containing the datagram. It would be nice have a test that makes sure that we don't at some point reject 0-RTT, but still process DATAGRAM frames arriving in 0-RTT packets.

Reply:
Oh I see, we have a misunderstanding here :)

What I was suggesting was calling `ReceiveMessage` on the server side, and making sure that it doesn't return the message sent by the client. Does that make sense?

Reply:
Thanks for being so patient :)

The `ReceiveMessage` returns the error in this testcase because the `EnableDatagrams` is not set, which is indeed unrelated to 0-RTT. 
So checking the tracer doesn't have the 0-RTT packet (means the server has dropped it) is enough here?  Sorry if I misunderstood your words again..

Reply:
Oh right, I missed that you're already calling `ReceiveMessage`. All good. Sorry!